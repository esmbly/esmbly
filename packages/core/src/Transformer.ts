import {
  AST,
  OutputFormat,
  Output,
  Transformer as TransformerInterface,
} from '@esmbly/types';

export abstract class Transformer implements TransformerInterface {
  public abstract readonly outputFormats: OutputFormat[];
  public abstract run(astArray: AST[]): AST[];
  public hasOutputFormat(output: Output[]): boolean {
    return output.some(
      (out: Output): boolean => {
        const format = typeof out === 'string' ? out : out.format;
        return this.outputFormats.includes(format as OutputFormat);
      },
    );
  }
}
