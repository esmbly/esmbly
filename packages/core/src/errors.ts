export const MissingConfig = (): string => {
  return `Please provide a valid configuration object.
    {
      input: File[];
      transformers: Transformer[];
      output: Output[];
    }
  `;
};

export const NoInput = (): string => {
  return 'Please specify at least one file to transform.';
};

export const NoTransformers = (): string => {
  return 'Please specify at least one transformer.';
};

export const NoOutput = (): string => {
  return 'Please specify at least one output format.';
};
