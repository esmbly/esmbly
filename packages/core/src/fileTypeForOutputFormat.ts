import { FileType, Format } from '@esmbly/types';

export function fileTypeForOutputFormat(format: Format): FileType {
  switch (format) {
    case Format.Asm:
      return FileType.Asm;
    case Format.Flow:
      return FileType.JavaScript;
    case Format.TypeScript:
      return FileType.TypeScript;
    case Format.Wat:
      return FileType.Wat;
    case Format.WebAssembly:
      return FileType.WebAssembly;
    case Format.AssemblyScript:
      return FileType.TypeScript;
    default:
      throw new Error(`Output format: ${format} is not supported`);
  }
}
