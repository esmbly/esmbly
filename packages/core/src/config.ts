import { RunConfig, Transformer, OutputFormat, Output } from '@esmbly/types';
import * as errors from './errors';

export function validateRunConfig(config: RunConfig): void {
  if (!config || !config.input || !config.transformers || !config.output) {
    throw new Error(errors.MissingConfig());
  }
  if (config.input.length < 1) {
    throw new Error(errors.NoInput());
  }
  if (config.transformers.length < 1) {
    throw new Error(errors.NoTransformers());
  }
  if (config.output.length < 1) {
    throw new Error(errors.NoOutput());
  }
  config.transformers.forEach((transformer: Transformer) => {
    if (typeof transformer.run !== 'function') {
      throw new Error(errors.InvalidTransformer(transformer));
    }
  });
  config.output.forEach((out: Output) => {
    let format = typeof out === 'string' ? out : out.format;
    if (format in OutputFormat === false) {
      throw new Error(errors.InvalidOutput(format));
    }
  });
}
