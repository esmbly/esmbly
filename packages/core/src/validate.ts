import { Format, RunConfig, SyntaxTree, Transformer } from '@esmbly/types';
import * as errors from './errors';

export function validateConfig(config?: RunConfig): void {
  if (!config || !config.input || !config.transformers || !config.output) {
    throw new Error(errors.MissingConfig());
  }
  if (config.input.length < 1) {
    throw new Error(errors.NoInput());
  }
  if (config.transformers.length < 1) {
    throw new Error(errors.NoTransformers());
  }
  if (config.output.length < 1) {
    throw new Error(errors.NoOutput());
  }
}

export function validateInputFormat(
  trees: SyntaxTree[],
  transformer: Transformer,
): void {
  trees.forEach((tree: SyntaxTree) => {
    if (transformer.format.input === Format.Any) {
      return;
    }
    if (tree.format !== transformer.format.input) {
      const { name } = transformer;
      throw new Error(
        `Transformer: ${name} does not support input format ${tree.format}`,
      );
    }
  });
}
