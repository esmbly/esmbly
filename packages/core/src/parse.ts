import * as parser from '@babel/parser';
import { File, Format, Output, SyntaxTree, Transformer } from '@esmbly/types';
import recast from 'recast';
import { fileTypeForOutputFormat } from './fileTypeForOutputFormat';

export function parse(files: File[], transformer: Transformer): SyntaxTree[] {
  return files.map(
    (file: File): SyntaxTree => {
      try {
        const tree = recast.parse(file.content.toString(), {
          parser: transformer.parser || {
            parse(source: string) {
              return parser.parse(source, {
                plugins: transformer.parserPlugins || [],
                sourceType: 'module',
              });
            },
          },
        });
        return {
          format: transformer.format.input,
          represents: file,
          setFormat(format: Format): void {
            this.format = format;
            this.represents.type = fileTypeForOutputFormat(format);
          },
          toCode(): string {
            return recast.print(tree).code;
          },
          toFile(output: Output, content?: string | Buffer): File {
            return {
              ...this.represents,
              content: content || this.toCode(),
              outputOptions: output,
            };
          },
          tree,
        };
      } catch (err) {
        const filePath = `${file.dir}/${file.name}`;
        throw new Error(
          `Error while parsing file: ${filePath}. ${err.toString()}`,
        );
      }
    },
  );
}
