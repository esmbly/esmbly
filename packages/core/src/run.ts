import { validateRunConfig } from './config';
import parser from '@esmbly/parser';
import { RunConfig, File } from '@esmbly/types';

export default async function run({
  input,
  transformers,
  output,
}: RunConfig): Promise<File[]> {
  validateRunConfig({ input, transformers, output });
  const outputFiles = [];
  let astArray = parser.parse(input);
  for (const transformer of transformers) {
    astArray = await transformer.run(astArray);
    for (const ast of astArray) {
      if (transformer.hasOutputFormat(output)) {
        outputFiles.push(ast.toFile());
      }
    }
  }
  return outputFiles;
}
