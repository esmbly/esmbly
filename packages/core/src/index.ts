import run from './run';
import { validateRunConfig } from './config';

export default { run, validateRunConfig };
export { Transformer } from './Transformer';
