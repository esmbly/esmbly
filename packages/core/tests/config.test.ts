import { Output, Transformer } from '@esmbly/types';
import { validateRunConfig } from '../src/config';
import mockConfig from './__fixtures__/config';

describe('validateRunConfigRunConfig', () => {
  it('throws an error if config object is missing', () => {
    expect(() => {
      validateRunConfig();
    }).toThrowErrorMatchingSnapshot();
  });

  it('throws an error if no files are provided', () => {
    expect(() => {
      validateRunConfig({
        ...mockConfig,
        input: [],
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it('throws an error if no transformers are provided', () => {
    expect(() => {
      validateRunConfig({
        ...mockConfig,
        transformers: [],
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it('throws an error if no output formats are provided', () => {
    expect(() => {
      validateRunConfig({
        ...mockConfig,
        output: [],
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it('throws an error if a transformer.transform is not a function', () => {
    expect(() => {
      validateRunConfig({
        ...mockConfig,
        transformers: [{}] as Transformer[],
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it('throws an error if an output format is not supported', () => {
    expect(() => {
      validateRunConfig({
        ...mockConfig,
        output: [{ format: 'Unknown type' } as unknown] as Output[],
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it('does not throw an error when passed a correct config object', () => {
    expect(() => {
      validateRunConfig(mockConfig);
    }).not.toThrow();
  });
});
