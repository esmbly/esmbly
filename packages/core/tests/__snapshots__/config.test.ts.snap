// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRunConfigRunConfig throws an error if a output format is not supported 1`] = `"Output format: \\"Unknown type\\" is not supported."`;

exports[`validateRunConfigRunConfig throws an error if a transformer is not a function 1`] = `"[object Object].run() is not a function."`;

exports[`validateRunConfigRunConfig throws an error if config object is missing 1`] = `
"Please provide a valid configuration object.
    {
      input: File[];
      transformers: Transformer[];
      output: Output[];
    }
  "
`;

exports[`validateRunConfigRunConfig throws an error if no files are provided 1`] = `"Please specify at least one file to transform."`;

exports[`validateRunConfigRunConfig throws an error if no output formats are provided 1`] = `"Please specify at least one output format."`;

exports[`validateRunConfigRunConfig throws an error if no transformers are provided 1`] = `"Please specify at least one transformer."`;
